### ISSUE TITLE
Convolutions don't work with NanGuardMode

### ISSUE THREAD

**Issue by matthias-k** [Theano/Theano#3910]
While trying to debug some NaNs in my function, I came across the problem that I can't use NanGuardMode on a function involving convolutions:

```python
import numpy as np
import theano
import theano.tensor as T
from theano.compile.nanguardmode import NanGuardMode

input = T.tensor4('input')
W = theano.shared(np.random.randn(10,3,10,10))
conv = T.nnet.conv2d(input, W)
loss = conv.sum()
f1 = theano.function([input], loss)
f2 = theano.function([input], loss,
                    mode=NanGuardMode())
```
While f1 runs fine, f2 raises a `NotImplementedError`:

```python
NotImplementedError                       Traceback (most recent call last)
/usr/local/lib/python3.4/dist-packages/theano/gof/link.py in f()
   1011                 try:
-> 1012                     wrapper(i, node, *thunks)
   1013                 except Exception:

/usr/local/lib/python3.4/dist-packages/theano/compile/nanguardmode.py in nan_check(i, node, fn)
    322                     do_check_on(x[0], node, fn, True)
--> 323             fn()
    324             outputs = fn.outputs

/usr/local/lib/python3.4/dist-packages/theano/gof/op.py in rval(p, i, o, n)
    882             def rval(p=p, i=node_input_storage, o=node_output_storage, n=node):
--> 883                 r = p(n, [x[0] for x in i], o)
    884                 for o in node.outputs:

/usr/local/lib/python3.4/dist-packages/theano/tensor/nnet/abstract_conv.py in perform(self, node, inp, out_)
    306         raise NotImplementedError(
--> 307             'AbstractConv2d theano optimization failed. '
    308             'Did you exclude both "conv_dnn" and "conv_gemm" from '

NotImplementedError: AbstractConv2d theano optimization failed. Did you exclude both "conv_dnn" and "conv_gemm" from the optimizer? Is cudnn available and does the GPU support it?

During handling of the above exception, another exception occurred:

NotImplementedError                       Traceback (most recent call last)
<ipython-input-20-0f706dd28e6e> in <module>()
----> 1 f2(input_data)

/usr/local/lib/python3.4/dist-packages/theano/compile/function_module.py in __call__(self, *args, **kwargs)
    857         t0_fn = time.time()
    858         try:
--> 859             outputs = self.fn()
    860         except Exception:
    861             if hasattr(self.fn, 'position_of_error'):

/usr/local/lib/python3.4/dist-packages/theano/gof/link.py in f()
   1012                     wrapper(i, node, *thunks)
   1013                 except Exception:
-> 1014                     raise_with_op(node, *thunks)
   1015         f.thunk_groups = thunk_groups
   1016 

/usr/local/lib/python3.4/dist-packages/theano/gof/link.py in raise_with_op(node, thunk, exc_info, storage_map)
    312         # extra long error message in that case.
    313         pass
--> 314     reraise(exc_type, exc_value, exc_trace)
    315 
    316 

/usr/local/lib/python3.4/dist-packages/six.py in reraise(tp, value, tb)
    683             value = tp()
    684         if value.__traceback__ is not tb:
--> 685             raise value.with_traceback(tb)
    686         raise value
    687 

/usr/local/lib/python3.4/dist-packages/theano/gof/link.py in f()
   1010             for i, (thunks, node) in enumerate(izip(thunk_groups, order)):
   1011                 try:
-> 1012                     wrapper(i, node, *thunks)
   1013                 except Exception:
   1014                     raise_with_op(node, *thunks)

/usr/local/lib/python3.4/dist-packages/theano/compile/nanguardmode.py in nan_check(i, node, fn)
    321                 if var.owner is not None:
    322                     do_check_on(x[0], node, fn, True)
--> 323             fn()
    324             outputs = fn.outputs
    325             for x in outputs:

/usr/local/lib/python3.4/dist-packages/theano/gof/op.py in rval(p, i, o, n)
    881             # default arguments are stored in the closure of `rval`
    882             def rval(p=p, i=node_input_storage, o=node_output_storage, n=node):
--> 883                 r = p(n, [x[0] for x in i], o)
    884                 for o in node.outputs:
    885                     compute_map[o][0] = True

/usr/local/lib/python3.4/dist-packages/theano/tensor/nnet/abstract_conv.py in perform(self, node, inp, out_)
    305     def perform(self, node, inp, out_):
    306         raise NotImplementedError(
--> 307             'AbstractConv2d theano optimization failed. '
    308             'Did you exclude both "conv_dnn" and "conv_gemm" from '
    309             'the optimizer? Is cudnn available and does the GPU support it?')

NotImplementedError: AbstractConv2d theano optimization failed. Did you exclude both "conv_dnn" and "conv_gemm" from the optimizer? Is cudnn available and does the GPU support it?
Apply node that caused the error: AbstractConv2d{border_mode='valid', subsample=(1, 1), filter_flip=True, imshp=None, kshp=None}(input, <TensorType(float64, 4D)>)
Toposort index: 0
Inputs types: [TensorType(float64, 4D), TensorType(float64, 4D)]
Inputs shapes: [(1, 3, 100, 100), (10, 3, 10, 10)]
Inputs strides: [(9223372036854775807, 80000, 800, 8), (2400, 800, 80, 8)]
Inputs values: ['not shown', 'not shown']
Outputs clients: [[Sum{acc_dtype=float64}(AbstractConv2d{border_mode='valid', subsample=(1, 1), filter_flip=True, imshp=None, kshp=None}.0)]]

Backtrace when the node is created:
  File "<ipython-input-15-d9d7dcba8686>", line 1, in <module>
    conv = T.nnet.conv2d(input, W)

HINT: Use the Theano flag 'exception_verbosity=high' for a debugprint and storage map footprint of this apply node.
```

The Exception occurs on a normal theano installation without any additional configuration (especially, I did not activate the GPU).

**Comment by nouiz** [comment]
Try to update Theano, I did a related fix that could help this case.

**Comment by matthias-k** [comment]
I tested again with the latest master (fdfbab37146ee475b3fd17d8d104fb09bf3a8d5c). At least it is only one exception now:

```python
Traceback (most recent call last):
  File "test_nanguardmode.py", line 17, in <module>
    print(f2(input_data))
  File "/gpfs01/bethge/home/mkuemmerer/envs/theano_bug/local/lib/python2.7/site-packages/Theano-0.8.0.dev0-py2.7.egg/theano/compile/function_module.py", line 859, in __call__
    outputs = self.fn()
  File "/gpfs01/bethge/home/mkuemmerer/envs/theano_bug/local/lib/python2.7/site-packages/Theano-0.8.0.dev0-py2.7.egg/theano/gof/link.py", line 1014, in f
    raise_with_op(node, *thunks)
  File "/gpfs01/bethge/home/mkuemmerer/envs/theano_bug/local/lib/python2.7/site-packages/Theano-0.8.0.dev0-py2.7.egg/theano/gof/link.py", line 314, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "/gpfs01/bethge/home/mkuemmerer/envs/theano_bug/local/lib/python2.7/site-packages/Theano-0.8.0.dev0-py2.7.egg/theano/gof/link.py", line 1012, in f
    wrapper(i, node, *thunks)
  File "/gpfs01/bethge/home/mkuemmerer/envs/theano_bug/local/lib/python2.7/site-packages/Theano-0.8.0.dev0-py2.7.egg/theano/compile/nanguardmode.py", line 323, in nan_check
    fn()
  File "/gpfs01/bethge/home/mkuemmerer/envs/theano_bug/local/lib/python2.7/site-packages/Theano-0.8.0.dev0-py2.7.egg/theano/gof/op.py", line 908, in rval
    r = p(n, [x[0] for x in i], o)
  File "/gpfs01/bethge/home/mkuemmerer/envs/theano_bug/local/lib/python2.7/site-packages/Theano-0.8.0.dev0-py2.7.egg/theano/tensor/nnet/abstract_conv.py", line 307, in perform
    'AbstractConv2d theano optimization failed. '
NotImplementedError: AbstractConv2d theano optimization failed. Did you exclude both "conv_dnn" and "conv_gemm" from the optimizer? Is cudnn available and does the GPU support it?
Apply node that caused the error: AbstractConv2d{border_mode='valid', subsample=(1, 1), filter_flip=True, imshp=None, kshp=None}(input, <TensorType(float64, 4D)>)
Toposort index: 0
Inputs types: [TensorType(float64, 4D), TensorType(float64, 4D)]
Inputs shapes: [(1, 3, 100, 100), (10, 3, 10, 10)]
Inputs strides: [(240000, 80000, 800, 8), (2400, 800, 80, 8)]
Inputs values: ['not shown', 'not shown']
Outputs clients: [[Sum{acc_dtype=float64}(AbstractConv2d{border_mode='valid', subsample=(1, 1), filter_flip=True, imshp=None, kshp=None}.0)]]

Backtrace when the node is created:
  File "test_nanguardmode.py", line 8, in <module>
    conv = T.nnet.conv2d(input, W)

```

**Comment by nouiz** [comment]
Do you have a c compiler available? can you print theano.config.cxx?

**Comment by matthias-k** [comment]
Yes, I do. `theano.config.cxx` is `'g++'`, which is available, too.

**Comment by lamblin** [comment]
So, calling `NanGuardMode()` does not include any optimizations. I don't remember if it was by design, or a mistake.
If you pass `mode=NanGuardMode(optimizer='fast_compile')` or `mode=NanGuardMode(optimizer='fast_run')`, then it seems to work.

I think that the default value for the optimizer should be `'default'` instead of `None` to prevent those issues. Does someone remember a rationale for disabling all optimizations by default?

**Comment by nouiz** [comment]
I think you are right. It was an oversight or an interface change that didn't got in NanGuardMode. I did a quick PR that fix this: https://github.com/Theano/Theano/pull/3968

**Comment by lamblin** [comment]
Should be fixed by #3968
